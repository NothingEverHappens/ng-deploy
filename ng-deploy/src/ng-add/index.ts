import { Tree } from '@angular-devkit/schematics';
import { from, Observable } from 'rxjs';

import { Project } from '../shared/types';

import { ngAdd } from './ng-add';
import { listProjects, projectPrompt } from '../shared/utils';

interface DeployOptions {
    project: string;
}

// You don't have to export the function as default. You can also have more than one rule factory
// per file.
export function ngDeploy({project}: DeployOptions): (host: Tree) => Observable<Tree> {
    return (host: Tree): Observable<Tree> => {

        return from<Promise<Tree>>(
            listProjects().then((projects: Project[]) => {
                return projectPrompt(projects).then(({firebaseProject}: any) => {
                    return ngAdd(host, {firebaseProject, project})
                });
            })
        );
    };
}
