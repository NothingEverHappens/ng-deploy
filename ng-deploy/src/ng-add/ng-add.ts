import { SchematicsException, Tree } from '@angular-devkit/schematics';
import { FirebaseJSON, FirebaseRc } from '../shared/types';
import { experimental, JsonParseMode, parseJson } from '@angular-devkit/core';

const stringifyFormatted = (obj: any) => JSON.stringify(obj, null, 2);

function emptyFirebaseJson() {
    return {
        hosting: []
    }
}

function emptyFirebaseRc() {
    return {
        targets: {}
    };
}


function generateHostingConfig(project: string, dist: string) {
    return {
        target: project,
        public: dist,
        ignore: ['firebase.json', '**/.*', '**/node_modules/**'],
        rewrites: [
            {
                source: '**',
                destination: '/index.html'
            }
        ]
    }
}

function generateFirebaseJson(tree: Tree, path: string, project: string, dist: string) {
    const firebaseJson: FirebaseJSON = tree.exists(path) ? JSON.parse(tree.read(path)!.toString()) : emptyFirebaseJson();

    if (firebaseJson.hosting.find(config => config.target === project)) {
        throw new SchematicsException(`Target ${project} already exists in firebase.json`);
    }

    firebaseJson.hosting.push(generateHostingConfig(project, dist));

    overwriteIfExists(tree, path, stringifyFormatted(firebaseJson));
}


function generateFirebaseRcTarget(firebaseProject: string, project: string) {
    return {
        "hosting": {
            [project]: [
                // TODO(kirjs): Generally site name is consistent with the project name, but there are edge cases.
                firebaseProject
            ]
        }
    };
}

function generateFirebaseRc(tree: Tree, path: string, firebaseProject: string, project: string) {
    const firebaseRc: FirebaseRc = tree.exists(path) ? JSON.parse(tree.read(path)!.toString()) : emptyFirebaseRc();

    if (firebaseProject in firebaseRc.targets) {
        throw new SchematicsException(`Firebase project ${firebaseProject} already defined in .firebaserc`);
    }

    firebaseRc.targets[firebaseProject] = generateFirebaseRcTarget(firebaseProject, project);

    overwriteIfExists(tree, path, stringifyFormatted(firebaseRc));
}

const overwriteIfExists = (tree: Tree, path: string, content: string) => {
    if (tree.exists(path)) tree.overwrite(path, content);
    else tree.create(path, content);
};

function getWorkspace(
    host: Tree,
): { path: string, workspace: experimental.workspace.WorkspaceSchema } {
    const possibleFiles = ['/angular.json', '/.angular.json'];
    const path = possibleFiles.filter(path => host.exists(path))[0];

    const configBuffer = host.read(path);
    if (configBuffer === null) {
        throw new SchematicsException(`Could not find angular.json`);
    }
    const content = configBuffer.toString();

    let workspace: experimental.workspace.WorkspaceSchema;
    try {
        workspace = parseJson(
            content,
            JsonParseMode.Loose,
        ) as {} as experimental.workspace.WorkspaceSchema;
    } catch (e) {
        throw new SchematicsException(`Could not parse angular.json: ` + e.message);
    }

    return {
        path,
        workspace,
    };
}


interface NgAddOptions {
    firebaseProject: string;
    project: string;
}

export function ngAdd(tree: Tree, options: NgAddOptions) {
    if (!options.project) {
        throw new SchematicsException('Option "project" is required');
    }

    // TODO(kirjs): Uncomment
    // const ngDeploy = {
    //   type: NodeDependencyType.Dev,
    //   name: 'ng-deploy',
    //   version: '../ng-deploy'
    // };
    //  TODO: Research if there is a better alternative
    // addPackageJsonDependency(tree, ngDeploy);

    // _context.addTask(new NodePackageInstallTask());

    const {path: workspacePath, workspace} = getWorkspace(tree);
    const project = workspace.projects[options.project];
    if (!project) {
        throw new SchematicsException('Project is not defined in this workspace');
    }

    if (project.projectType !== 'application') {
        throw new SchematicsException(`Deploy requires a project type of "application" in angular.json`);
    }

    if (!project.architect || !project.architect.build || !project.architect.build.options || !project.architect.build.options.outputPath) {
        throw new SchematicsException(`Cannot read the output path (architect.build.options.outputPath) of project "${options.project}" in angular.json`);
    }

    const outputPath = project.architect.build.options.outputPath;

    project.architect['deploy'] = {
        builder: 'ng-deploy:deploy',
        options: {}
    };

    tree.overwrite(workspacePath, JSON.stringify(workspace, null, 2));
    generateFirebaseJson(tree, '/firebase.json', options.project, outputPath);
    generateFirebaseRc(tree, '/.firebaserc', options.firebaseProject, options.project);
    return tree;
}
