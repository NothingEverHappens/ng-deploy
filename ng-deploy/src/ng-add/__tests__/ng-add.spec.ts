import { Tree } from '@angular-devkit/schematics';
import { ngAdd } from '../ng-add';


const PROJECT_NAME = 'pie-ka-chu';
const PROJECT_ROOT = 'pirojok';
const FIREBASE_PROJECT = 'pirojok-111e3';

const OTHER_PROJECT_NAME = 'pi-catch-you';
const OTHER_FIREBASE_PROJECT_NAME = 'bi-catch-you-77e7e';

describe('ng-add', () => {
    describe('generating files', () => {
        let tree: Tree;

        beforeEach(() => {
            tree = Tree.empty();
            tree.create('angular.json', JSON.stringify(generateAngularJson()));
        });

        it('generates new files if starting from scratch', async () => {
            const result = ngAdd(tree, {firebaseProject: FIREBASE_PROJECT, project: PROJECT_NAME});
            expect(result.read('firebase.json')!.toString()).toMatchSnapshot();
            expect(result.read('.firebaserc')!.toString()).toMatchSnapshot();
            expect(result.read('angular.json')!.toString()).toMatchSnapshot();
        });

        it('uses default project', async () => {
            const result = ngAdd(tree, {firebaseProject: FIREBASE_PROJECT});
            expect(result.read('firebase.json')!.toString()).toMatchSnapshot();
            expect(result.read('.firebaserc')!.toString()).toMatchSnapshot();
            expect(result.read('angular.json')!.toString()).toMatchSnapshot();
        });

        it('overrides existing files', async () => {
            const tempTree = ngAdd(tree, {firebaseProject: FIREBASE_PROJECT, project: PROJECT_NAME});
            const result = ngAdd(tempTree, {firebaseProject: OTHER_FIREBASE_PROJECT_NAME, project: OTHER_PROJECT_NAME});
            expect(result.read('firebase.json')!.toString()).toMatchSnapshot();
            expect(result.read('.firebaserc')!.toString()).toMatchSnapshot();
            expect(result.read('angular.json')!.toString()).toMatchSnapshot();
        });
    });

    describe('error handling', () => {
        it('fails if project not defined', () => {
            const tree = Tree.empty();
            const angularJSON = generateAngularJson();
            delete angularJSON.defaultProject;
            tree.create('angular.json', JSON.stringify(angularJSON));
            expect(() => ngAdd(tree, {
                firebaseProject: FIREBASE_PROJECT,
                project: ''
            })).toThrowError(/No project selected and no default project in the workspace/);
        });

        it('Should throw if angular.json not found', async () => {
            expect(() => ngAdd(Tree.empty(), {
                firebaseProject: FIREBASE_PROJECT,
                project: PROJECT_NAME
            })).toThrowError(/Could not find angular.json/)
        });

        it('Should throw if angular.json  can not be parsed', async () => {
            const tree = Tree.empty();
            tree.create('angular.json', 'hi');
            expect(() => ngAdd(tree, {
                firebaseProject: FIREBASE_PROJECT,
                project: PROJECT_NAME
            })).toThrowError(/Could not parse angular.json/)
        });

        it('Should throw if specified project does not exist ', async () => {
            const tree = Tree.empty();
            tree.create('angular.json', JSON.stringify({projects: {}}));
            expect(() => ngAdd(tree, {
                firebaseProject: FIREBASE_PROJECT,
                project: PROJECT_NAME
            })).toThrowError(/Project is not defined in this workspace/)
        });

        it('Should throw if specified project is not application', async () => {
            const tree = Tree.empty();
            tree.create('angular.json', JSON.stringify({projects: {[PROJECT_NAME]: {projectType: 'pokemon'}}}));
            expect(() => ngAdd(tree, {
                firebaseProject: FIREBASE_PROJECT,
                project: PROJECT_NAME
            })).toThrowError(/Deploy requires a project type of "application"/)
        });

        it('Should throw if app does not have architect configured', async () => {
            const tree = Tree.empty();
            tree.create('angular.json', JSON.stringify({projects: {[PROJECT_NAME]: {projectType: 'application'}}}));
            expect(() => ngAdd(tree, {
                firebaseProject: FIREBASE_PROJECT,
                project: PROJECT_NAME
            })).toThrowError(/Cannot read the output path/)
        });

        it('Should throw if firebase.json has the project already', async () => {
            const tree = Tree.empty();
            tree.create('angular.json', JSON.stringify(generateAngularJson()));
            const tempTree = ngAdd(tree, {firebaseProject: FIREBASE_PROJECT, project: PROJECT_NAME});

            expect(() => ngAdd(tempTree, {
                firebaseProject: FIREBASE_PROJECT,
                project: PROJECT_NAME
            })).toThrowError(/already exists in firebase.json/);
        });

        it('Should throw if firebase.json is broken', async () => {
            const tree = Tree.empty();
            tree.create('angular.json', JSON.stringify(generateAngularJson()));
            tree.create('firebase.json', 'I\'m broken ðŸ˜”');
            expect(() => ngAdd(tree, {
                firebaseProject: FIREBASE_PROJECT,
                project: PROJECT_NAME
            })).toThrowError(/firebase.json: Unexpected token/);
        });

        it('Should throw if .firebaserc is broken', async () => {
            const tree = Tree.empty();
            tree.create('angular.json', JSON.stringify(generateAngularJson()));
            tree.create('.firebaserc', 'I\'m broken ðŸ˜”');
            expect(() => ngAdd(tree, {
                firebaseProject: FIREBASE_PROJECT,
                project: PROJECT_NAME
            })).toThrowError(/.firebaserc: Unexpected token/);
        });

        it('Should throw if firebase.json has the project already', async () => {
            const tree = Tree.empty();
            tree.create('angular.json', JSON.stringify(generateAngularJson()));
            const tempTree = ngAdd(tree, {firebaseProject: FIREBASE_PROJECT, project: PROJECT_NAME});

            expect(() => ngAdd(tempTree, {
                firebaseProject: FIREBASE_PROJECT,
                project: OTHER_PROJECT_NAME
            })).toThrowError(/ already defined in .firebaserc/);
        });

        it('Should throw if firebase.json is broken', async () => {
            const tree = Tree.empty();
            tree.create('angular.json', JSON.stringify(generateAngularJson()));

            const tempTree = ngAdd(tree, {firebaseProject: FIREBASE_PROJECT, project: PROJECT_NAME});

            expect(() => ngAdd(tempTree, {
                firebaseProject: FIREBASE_PROJECT,
                project: OTHER_PROJECT_NAME
            })).toThrowError(/ already defined in .firebaserc/);
        });
    });
});


function generateAngularJson() {
    return {
        defaultProject: PROJECT_NAME,
        projects: {
            [PROJECT_NAME]: {
                projectType: 'application',
                root: PROJECT_ROOT,
                architect: {
                    build: {
                        options: {
                            outputPath: 'dist/ikachu'
                        }
                    },
                }
            },
            [OTHER_PROJECT_NAME]: {
                projectType: 'application',
                root: PROJECT_ROOT,
                architect: {
                    build: {
                        options: {
                            outputPath: 'dist/ikachu'
                        }
                    },
                }
            }
        }
    };
}
