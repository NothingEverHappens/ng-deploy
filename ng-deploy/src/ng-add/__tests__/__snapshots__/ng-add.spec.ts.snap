// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ng-add generating files generates new files if starting from scratch 1`] = `
"{
  \\"hosting\\": [
    {
      \\"target\\": \\"pie-ka-chu\\",
      \\"public\\": \\"dist/ikachu\\",
      \\"ignore\\": [
        \\"firebase.json\\",
        \\"**/.*\\",
        \\"**/node_modules/**\\"
      ],
      \\"rewrites\\": [
        {
          \\"source\\": \\"**\\",
          \\"destination\\": \\"/index.html\\"
        }
      ]
    }
  ]
}"
`;

exports[`ng-add generating files generates new files if starting from scratch 2`] = `
"{
  \\"targets\\": {
    \\"pirojok-111e3\\": {
      \\"hosting\\": {
        \\"pie-ka-chu\\": [
          \\"pirojok-111e3\\"
        ]
      }
    }
  }
}"
`;

exports[`ng-add generating files generates new files if starting from scratch 3`] = `
"{
  \\"defaultProject\\": \\"pie-ka-chu\\",
  \\"projects\\": {
    \\"pie-ka-chu\\": {
      \\"projectType\\": \\"application\\",
      \\"root\\": \\"pirojok\\",
      \\"architect\\": {
        \\"build\\": {
          \\"options\\": {
            \\"outputPath\\": \\"dist/ikachu\\"
          }
        },
        \\"deploy\\": {
          \\"builder\\": \\"ng-deploy:deploy\\",
          \\"options\\": {}
        }
      }
    },
    \\"pi-catch-you\\": {
      \\"projectType\\": \\"application\\",
      \\"root\\": \\"pirojok\\",
      \\"architect\\": {
        \\"build\\": {
          \\"options\\": {
            \\"outputPath\\": \\"dist/ikachu\\"
          }
        }
      }
    }
  }
}"
`;

exports[`ng-add generating files overrides existing files 1`] = `
"{
  \\"hosting\\": [
    {
      \\"target\\": \\"pie-ka-chu\\",
      \\"public\\": \\"dist/ikachu\\",
      \\"ignore\\": [
        \\"firebase.json\\",
        \\"**/.*\\",
        \\"**/node_modules/**\\"
      ],
      \\"rewrites\\": [
        {
          \\"source\\": \\"**\\",
          \\"destination\\": \\"/index.html\\"
        }
      ]
    },
    {
      \\"target\\": \\"pi-catch-you\\",
      \\"public\\": \\"dist/ikachu\\",
      \\"ignore\\": [
        \\"firebase.json\\",
        \\"**/.*\\",
        \\"**/node_modules/**\\"
      ],
      \\"rewrites\\": [
        {
          \\"source\\": \\"**\\",
          \\"destination\\": \\"/index.html\\"
        }
      ]
    }
  ]
}"
`;

exports[`ng-add generating files overrides existing files 2`] = `
"{
  \\"targets\\": {
    \\"pirojok-111e3\\": {
      \\"hosting\\": {
        \\"pie-ka-chu\\": [
          \\"pirojok-111e3\\"
        ]
      }
    },
    \\"bi-catch-you-77e7e\\": {
      \\"hosting\\": {
        \\"pi-catch-you\\": [
          \\"bi-catch-you-77e7e\\"
        ]
      }
    }
  }
}"
`;

exports[`ng-add generating files overrides existing files 3`] = `
"{
  \\"defaultProject\\": \\"pie-ka-chu\\",
  \\"projects\\": {
    \\"pie-ka-chu\\": {
      \\"projectType\\": \\"application\\",
      \\"root\\": \\"pirojok\\",
      \\"architect\\": {
        \\"build\\": {
          \\"options\\": {
            \\"outputPath\\": \\"dist/ikachu\\"
          }
        },
        \\"deploy\\": {
          \\"builder\\": \\"ng-deploy:deploy\\",
          \\"options\\": {}
        }
      }
    },
    \\"pi-catch-you\\": {
      \\"projectType\\": \\"application\\",
      \\"root\\": \\"pirojok\\",
      \\"architect\\": {
        \\"build\\": {
          \\"options\\": {
            \\"outputPath\\": \\"dist/ikachu\\"
          }
        },
        \\"deploy\\": {
          \\"builder\\": \\"ng-deploy:deploy\\",
          \\"options\\": {}
        }
      }
    }
  }
}"
`;

exports[`ng-add generating files uses default project 1`] = `
"{
  \\"hosting\\": [
    {
      \\"target\\": \\"pie-ka-chu\\",
      \\"public\\": \\"dist/ikachu\\",
      \\"ignore\\": [
        \\"firebase.json\\",
        \\"**/.*\\",
        \\"**/node_modules/**\\"
      ],
      \\"rewrites\\": [
        {
          \\"source\\": \\"**\\",
          \\"destination\\": \\"/index.html\\"
        }
      ]
    }
  ]
}"
`;

exports[`ng-add generating files uses default project 2`] = `
"{
  \\"targets\\": {
    \\"pirojok-111e3\\": {
      \\"hosting\\": {
        \\"pie-ka-chu\\": [
          \\"pirojok-111e3\\"
        ]
      }
    }
  }
}"
`;

exports[`ng-add generating files uses default project 3`] = `
"{
  \\"defaultProject\\": \\"pie-ka-chu\\",
  \\"projects\\": {
    \\"pie-ka-chu\\": {
      \\"projectType\\": \\"application\\",
      \\"root\\": \\"pirojok\\",
      \\"architect\\": {
        \\"build\\": {
          \\"options\\": {
            \\"outputPath\\": \\"dist/ikachu\\"
          }
        },
        \\"deploy\\": {
          \\"builder\\": \\"ng-deploy:deploy\\",
          \\"options\\": {}
        }
      }
    },
    \\"pi-catch-you\\": {
      \\"projectType\\": \\"application\\",
      \\"root\\": \\"pirojok\\",
      \\"architect\\": {
        \\"build\\": {
          \\"options\\": {
            \\"outputPath\\": \\"dist/ikachu\\"
          }
        }
      }
    }
  }
}"
`;
